File: next-js-boilerplate/src/api/client/auth/actions.ts
"use client";

import { loginServer } from "@/api/server/auth/login";
import { registerServer } from "@/api/server/auth/register";
import { logoutServer } from "@/api/server/auth/logout";
import { refreshTokenServer } from "@/api/server/auth/token";
import {
  verifyEmailServer,
  verifyEmailCodeServer,
  resendEmailCodeServer,
} from "@/api/server/auth/verify-email";
import { requestPasswordResetServer } from "@/api/server/auth/request-password-reset";
import { resetPasswordServer } from "@/api/server/auth/reset-password";
import { changePasswordServer } from "@/api/server/auth/change-password";
import { undoPasswordChangeServer } from "@/api/server/auth/undo-password-change";
import {
  enrollMfaServer,
  verifyMfaEnrollmentServer,
  disableMfaServer,
  resendLoginCodeServer,
} from "@/api/server/auth/mfa";
import type {
  EnrollMfaResult,
  VerifyMfaResult,
  DisableMfaResult,
} from "@/api/server/auth/mfa";

export function useAuthActions() {
  const login = async (email: string, password: string, timezone?: string) =>
    loginServer(email, password, timezone);

  const register = async (
    email: string,
    password: string,
    name?: string,
    timezone?: string,
  ) => registerServer(email, password, name, timezone);

  const logout = async () => logoutServer();

  const refreshToken = async () => refreshTokenServer();

  const verifyEmail = async (token: string) => verifyEmailServer(token);

  const verifyEmailCode = async (userId: string, code: string) =>
    verifyEmailCodeServer(userId, code);

  const resendEmailCode = async (userId: string, email: string) =>
    resendEmailCodeServer(userId, email);

  const requestPasswordReset = async (email: string) =>
    requestPasswordResetServer(email);

  const resetPassword = async (token: string, newPassword: string) =>
    resetPasswordServer(token, newPassword);

  const changePassword = async (currentPassword: string, newPassword: string) =>
    changePasswordServer(currentPassword, newPassword);

  const undoPasswordChange = async (token: string) =>
    undoPasswordChangeServer(token);

  const enrollMfa = async (): Promise<EnrollMfaResult> => enrollMfaServer();

  const verifyMfaEnrollment = async (code: string): Promise<VerifyMfaResult> =>
    verifyMfaEnrollmentServer(code);

  const disableMfa = async (code: string): Promise<DisableMfaResult> =>
    disableMfaServer(code);

  const resendLoginCode = async (mfaToken: string) =>
    resendLoginCodeServer(mfaToken);

  return {
    login,
    register,
    logout,
    refreshToken,
    verifyEmail,
    verifyEmailCode,
    resendEmailCode,
    requestPasswordReset,
    resetPassword,
    changePassword,
    undoPasswordChange,
    enrollMfa,
    verifyMfaEnrollment,
    disableMfa,
    resendLoginCode,
  };
}

File: flutter-boilerplate/lib/api/client/auth/actions.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../types/auth/auth_request_types.dart';
import '../../../types/auth/mfa_types.dart';
import '../../server/auth/login.dart';
import '../../server/auth/logout.dart';
import '../../server/auth/mfa.dart';
import '../../server/auth/register.dart';
import '../../server/auth/request_password_reset.dart';
import '../../server/auth/reset_password.dart';
import '../../server/auth/verify_email.dart';

final loginActionsProvider = Provider((ref) => LoginActions(ref));

class LoginActions {
  final Ref _ref;

  LoginActions(this._ref);

  Future<LoginResult> login(LoginRequest request) async {
    final server = _ref.read(loginServerProvider);
    return server.call(request);
  }

  Future<Map<String, dynamic>> verifyLoginMfa(
    String mfaToken,
    String code,
  ) async {
    final server = _ref.read(mfaServerProvider);
    return server.verifyLoginMfa(mfaToken, code);
  }

  Future<RegisterResponse> register(RegisterRequest request) async {
    final server = _ref.read(registerServerProvider);
    return server.call(request);
  }

  Future<void> logout() async {
    final server = _ref.read(logoutServerProvider);
    await server.call();
  }

  Future<void> requestPasswordReset(String email) async {
    final server = _ref.read(requestPasswordResetServerProvider);
    await server.call(email);
  }

  Future<void> resetPassword(String token, String password) async {
    final server = _ref.read(resetPasswordServerProvider);
    await server.call(token, password);
  }

  Future<void> verifyEmail(String token) async {
    final server = _ref.read(verifyEmailServerProvider);
    await server.call(token);
  }

  Future<void> verifyEmailCode(String userId, String code) async {
    final server = _ref.read(verifyEmailServerProvider);
    await server.verifyCode(userId, code);
  }

  Future<bool> resendEmailCode(String userId, String email) async {
    final server = _ref.read(verifyEmailServerProvider);
    return server.resendCode(userId, email);
  }

  Future<EnrollMfaResponse> enrollMfa() async {
    final server = _ref.read(mfaServerProvider);
    return server.enrollMfa();
  }

  Future<VerifyMfaResponse> verifyMfa(String code) async {
    final server = _ref.read(mfaServerProvider);
    return server.verifyMfa(code);
  }

  Future<bool> disableMfa(String code) async {
    final server = _ref.read(mfaServerProvider);
    return server.disableMfa(code);
  }

  Future<String> resendLoginCode(String mfaToken) async {
    final server = _ref.read(mfaServerProvider);
    return server.resendLoginCode(mfaToken);
  }
}
